### ISSUE TITLE
File comparison

### ISSUE THREAD

**Issue by kirawi** [helix-editor/helix#405]
I have a vague idea on how to do this, but it'll take me some time to familiarize myself with the UI code.

**Comment by zbindenren** [comment]
Maybe delta can help in that case: https://github.com/dandavison/delta

It is the best terminal diff viewer I ever used  and it is also written in rust.

**Comment by sudormrfbin** [comment]
Delta is a pager that acts on already diffed data (with `-` and `+` prepended to diffed lines like in the default git diff output) and simply applies syntax highlighting and better colors it. We will have to use [similar](https://github.com/mitsuhiko/similar) for this, which we already use in helix to `:reload` the file from disk.

**Comment by archseer** [comment]
@pascalkuthe mentioned he'll probably pick up the work on this in the future after git gutter is merged.

**Comment by pascalkuthe** [comment]
After #3890 is merged, the main blocker will be rendering the diffs side by side (as a form of virtual text).
I am working with @kirawi to land the necessary changes in the rendering system to make that happen.

**Comment by LifeWorks** [comment]
Is this feature available in the current version? I see it is open, so I assume I still can not diff two buffers?

**Comment by archseer** [comment]
Yes, open issues typically indicate that a feature hasn't been implemented yet.

**Comment by wh75er** [comment]
Hello! Would it be possible to use inline git diff in one buffer? It would be amazing

**Comment by stevenxxiu** [comment]
> Hello! Would it be possible to use inline git diff in one buffer? It would be amazing

Please open a separate issue for that.

**Comment by Dmole** [comment]
There are 2 uses cases I'd want helixdiff to handle:

- simple diff of 2 config files
- git conflictstyle diff3 which can use:

    vimdiff -f -d -c '4wincmd w | wincmd J' "$LOCAL" "$BASE" "$REMOTE" "$MERGED"

**Comment by DocteurFlashbios** [comment]
It would be really great to offer a more clear and advanced way to diff/merge files than vim. Either threeway merge is supported, but nothing is shown about who is who and I never managed to remember the "fingertricks"

**Comment by Dmole** [comment]
> It would be really great to offer a more clear and advanced way to diff/merge files than vim. Either threeway merge is supported, but nothing is shown about who is who and I never managed to remember the "fingertricks"

Is 
    :ls
or
    :set statusline=%n\ %f
not what you want?

**Comment by DocteurFlashbios** [comment]
> > It would be really great to offer a more clear and advanced way to diff/merge files than vim. Either threeway merge is supported, but nothing is shown about who is who and I never managed to remember the "fingertricks"
> 
> Is :ls or :set statusline=%n\ %f not what you want?

I thought about some vimdiff features like https://devhints.io/vim-diff
which are very usefull for merging files or branches

**Comment by gauteh** [comment]
Thanks for this great project! Just want to pitch in that diffing git revisions (fugitive functionality) and using helix as a difftool (to do reviewes or similar), fixing conflicts etc, would be really amazing. I don't really need a full git client (like committing etc, even logs), but I suppose those may be the easy parts.

**Comment by npupko** [comment]
Hey! Are there any ideas or requirements on how the first version should look like? Any specifics, like available commands, actions, etc.

What's the MVP for the scope of this PR? Should it be something comprehensive, or simple to be used, like:

<img width="3456" height="1972" alt="Image" src="https://github.com/user-attachments/assets/161096d4-883c-4c84-ba33-5ee13b05321a" />

I feel like this issue may became more prioritized due to increased AI assistance, such as Claude Code IDE integration, etc.

**Comment by brianmay** [comment]
For some sort of inspiration have a look vimdiff.

I would suggest we don't need to go all way though for the first version, e.g.

- limit to 2 files at once (vimdiff can load and compare any number of files).
- if syntax highlighting causes conflicts with diff highlighting, disable syntax highlighting.
- commands to automatically copy chunk/lines to/from one file to the other file (IIRC vimdiff only does complete chunks not lines, but might be wrong here).
- commands to jump to previous/next difference.

**Comment by webdev23** [comment]
Throwing a bunch of crates together and calling it post-modern was not enough, it seems, 4 years ago.

100+ upvote, 4 years and not a single soul on it, call it skill-issues. So many similar queries and PR strongly needed in developer workflow are simply ignored. We will have to wait for a new team, as this one is just inefficient, they don't read, don't listen what we say. At best you'll get hidden as off-topic if you slip and curve down their ego with obvious technical concerns. 

So, you could try old duct tape like this, it does open the full git log in a new tab, for quick copy/paste, and you could try to adapt to your need.

```
"A-g" = [
  ':insert-output git diff HEAD > /tmp/git_diff.txt',
  ':o "/tmp/git_diff.txt"',
  ':set-language diff'
]
```

Of course, there is news! Some minded engineers implemented that `%{buffer_name}` variable to call file name into commands. Great attempt!

Unfortunately, almost prophetically, using it just fails silently, but not always: (the following does not work for windy reasons). In other words don't use %{buffer_name} as of now, it's buggy.

```
"A-g" = [
  ':insert-output git diff HEAD %{buffer_name} > /tmp/git_diff.txt',
  ':o "/tmp/git_diff.txt"',
  ':set-language diff'
]
```
---

Now, more pragmatically, some programs not made by juniors are using alternate buffer and use the stderr to feed their buffer, and it allow some cool tricks:

Using Lazygit https://github.com/jesseduffield/lazygit

```
"C-g" = [
  ':sh rm -f /tmp/files2open',
  ':set mouse false',
  ':insert-output lazygit',
  ':redraw',
  ':set mouse true',
  ':open /tmp/lazygit_out.txt',
  'select_all',
  'split_selection_on_newline',
  'goto_file',
  ':buffer-close! /tmp/lazygit_out.txt',
  ':rla'
]
```

Additionally, to actually use Helix from Lazygit while pressing "e", we have to add something similar to this in the Lazygit config, also setting EDITOR="hx" in the shell environment.

It does allow to restore files, see diffs, and way more.

hx ~/.config/lazygit/config.yml
```
edit: "echo {{filename}} > /tmp/lazygit_out.txt && killall lazygit"
editInTerminal: true
```

This are just `git` diff, and them aren't the clearest nor the prettiest.

Of course, a plugin system is coming, so chances are that some version control tools will be made in the next decade. But making it better than Lazygit will definitely be hard. 

Good luck to all.

**Comment by mi2ebi** [comment]
mm i don't think insulting the maintainers was necessary there

**Comment by webdev23** [comment]
If you too you like downvoting truthful post (who cares seriously), try a lookup on this one about how Helix behave with `.gitignore`: https://github.com/helix-editor/helix/issues/12604#issuecomment-3148626353

**Comment by archseer** [comment]
> Throwing a bunch of crates together and calling it post-modern was not enough, it seems, 4 years ago.

@webdev23 It's easy to spew negativity from an anonymous pseudonym online. As previously mentioned in this thread, this project has been built by volunteers in their own spare time (not robots that are here to build software for you). [You aren't entitled to anything](https://gist.github.com/richhickey/1563cddea1002958f96e7ba9519972d9). Side by side diffing has been a highly desired feature for me too and [I fully understand your frustration](https://burntsushi.net/foss/#rudeness) (you have a genuine problem and a feature you need solves it) but I currently simply don't have the time to work on it at the moment, since it requires pretty significant changes to the rendering.

You say it's a "skill issue", well, why don't you show us how it's done and build it? I'll personally merge the PR if you implement the feature. We're all developers, when something is missing we can go and attempt to solve the problem. [You've so far only lodged complaints and haven't provided any patches](https://github.com/helix-editor/helix/issues?q=is%3Apr+is%3Aopen+author%3Awebdev23). Certainly, PRs will get you further than throwing negative comments at us. I personally find it pretty hurtful and demotivating every time I see yet another comment like this in my inbox which makes me even less likely to work on this feature.

**Comment by RoloEdits** [comment]
The way @webdev23 has presented themselves, they appear to have extensive knowledge in terminals, terminal rendering,  unicode, version control systems, operating systems, and various other specifications. If someone with this knowledge, and seemingly time, were to contribute, the project would be better as a whole. I think everyone would welcome such improvements! Hopefully we can see a PR soon to address some outstanding issues.

**Comment by webdev23** [comment]
`extensive knowledge in terminals, terminal rendering, unicode, version control systems, operating systems, and various other specifications.`

Yes, it's true, all that is not out of nowhere. However nothing is done, be humble. We are at a turning point between the bronze age and the real Iron age, in programming. We have more to do than arguing 4 years around 1970's file diffs. At least this is what i hope for you all.

Just push the green button here, Unix sockets: https://github.com/helix-editor/helix/pull/13896

![Image](https://github.com/user-attachments/assets/bee591b8-8b8e-4d0f-b56d-27a0cb9fbe06)

Thank you.

**Comment by archseer** [comment]
> We have more to do than arguing 4 years around 1970's file diffs

Looks like you ignored everything we said? We're all eagerly awaiting your first PR. Rather than issuing demands, collaborate!

https://docs.github.com/articles/creating-a-pull-request

**Comment by somephic** [comment]
As others have stated, I think looking at vimdiff's interface is smart and using libxdiff like it and git are using as well too.

However I *do* also think it'd be great if helix [implemented Kakoune's client-server model](https://github.com/helix-editor/helix/issues/312) so communication between splits, tmux windows, terminal windows or whatever could be kept in sync. Helix today can't [maximize splits](https://github.com/helix-editor/helix/issues/2262) which makes it a little less ergonomic if you want to temporarily expand one of the git-mergetool style 4 split panes for context while merging. You're stuck with it tiny. Hopefully your screen is wide enough.

Really, I think if the editor had event-based triggers, you could solve this with a helix config file in a more POSIX way than either building it in monolithically or waiting for the plugin panacea to save the day. If you run an unrelated helixdiff that sends N file_name args to a differ binary and updates N helix clients tied to one helix server instance, helix wouldn't have to do much of anything.

I use tmux dialog's popups to get a full `lf` file manager or `tig` git workflows. I just have to run `tmux send-keys ":o \"$file_name\"" Enter` to send outputs to helix in a more brittle way because it doesn't have remote-control capabilities like Kakoune.

Anyway, just some thoughts.

**Comment by ipochi** [comment]
If we just want to view the diff in the overlay on the hunks [i .e colored lines], then this works great.

```
[keys.normal.space]
v = ":sh git diff -U0 -- %{buffer_name} | awk -v l=%{cursor_line} '/^@@/ { if (found) exit; match($0, /\\+[0-9]+(,[0-9]+)?/); s=substr($0,RSTART+1,RLENGTH-1); split(s,a,\",\"); start=a[1]; count=(a[2] ? a[2] : 1); end=start+count-1; found=(l>=start && l<=end); if (found) print; next } found { print }'"
```

add this your helix config. If you are on the colored lines shown by helix then  "space + v" will show you the diff as an overlay

**Comment by ipochi** [comment]
you can either reach the hunk by [g or ]g then space + v or if you are on the colored line "space +v"
